### ISSUE TITLE
验证代码

### ISSUE THREAD

**Issue by xiaocui-iii** [yxlu-0102/MP-SENet#51]
您好，很感谢您发布了最新transformer块的代码，我想请问一下，后续还会发布验证最佳ckpt的代码吗，我想通过修改CMGAN的evaluation.py来验证最佳的ckpt，但是出了错误，如果您方便，可以回答一下嘛

`import numpy as np
from models import generator
from natsort import natsorted
import os
from tools.compute_metrics import compute_metrics
from utils import *
import torchaudio
import soundfile as sf
import argparse
import json
from env import AttrDict, build_env
from stfts import mag_phase_stft, mag_phase_istft

@torch.no_grad()
def enhance_one_track(
    model, audio_path, saved_dir, cut_len, h, save_tracks=False
):
    n_fft, hop_size, win_size = h.n_fft, h.hop_size, h.win_size
    compress_factor = h.compress_factor
    name = os.path.split(audio_path)[-1]
    noisy, sr = torchaudio.load(audio_path)
    assert sr == 16000
    noisy = noisy.cuda()

    # c = torch.sqrt(noisy.size(-1) / torch.sum((noisy**2.0), dim=-1))
    # noisy = torch.transpose(noisy, 0, 1)
    # noisy = torch.transpose(noisy * c, 0, 1)

    length = noisy.size(-1)
    frame_num = int(np.ceil(length / 100))
    padded_len = frame_num * 100
    padding_len = padded_len - length
    noisy = torch.cat([noisy, noisy[:, :padding_len]], dim=-1)
    if padded_len > cut_len:
        batch_size = int(np.ceil(padded_len / cut_len))
        while 100 % batch_size != 0:
            batch_size += 1
        noisy = torch.reshape(noisy, (batch_size, -1))

    noisy_mag, noisy_pha, noisy_com = mag_phase_stft(noisy, n_fft, hop_size, win_size, compress_factor, addeps=True)
    est_real, est_mag, est_com = model(noisy_mag, noisy_pha)
    est_audio = mag_phase_istft(est_mag, est_mag, n_fft, hop_size, win_size, compress_factor)
    est_audio = torch.flatten(est_audio)[:length].cpu().numpy()

    assert len(est_audio) == length
    if save_tracks:
        saved_path = os.path.join(saved_dir, name)
        sf.write(saved_path, est_audio, sr)

    return est_audio, length

def evaluation(model_path, noisy_dir, clean_dir, save_tracks, saved_dir, h):
    n_fft = 400
    model = generator.MPNet(h).cuda()
    model.load_state_dict((torch.load(model_path)), strict=False)
    model.eval()

    if not os.path.exists(saved_dir):
        os.mkdir(saved_dir)

    audio_list = os.listdir(noisy_dir)
    audio_list = natsorted(audio_list)
    num = len(audio_list)
    metrics_total = np.zeros(6)
    for audio in audio_list:
        noisy_path = os.path.join(noisy_dir, audio)
        clean_path = os.path.join(clean_dir, audio)
        est_audio, length = enhance_one_track(
            model, noisy_path, saved_dir, 16000 * 16, h, save_tracks
        )
        clean_audio, sr = sf.read(clean_path)
        assert sr == 16000
        metrics = compute_metrics(clean_audio, est_audio, sr, 0)
        metrics = np.array(metrics)
        metrics_total += metrics

    metrics_avg = metrics_total / num
    print(
        "pesq: ",
        metrics_avg[0],
        "csig: ",
        metrics_avg[1],
        "cbak: ",
        metrics_avg[2],
        "covl: ",
        metrics_avg[3],
        "ssnr: ",
        metrics_avg[4],
        "stoi: ",
        metrics_avg[5],
    )

parser = argparse.ArgumentParser()
parser.add_argument("--model_path", type=str, default='./best_ckpt/ckpt',
                    help="the path where the model is saved")
parser.add_argument("--test_dir", type=str, default='dir to your VCTK-DEMAND test dataset',
                    help="noisy tracks dir to be enhanced")
parser.add_argument("--save_tracks", type=str, default=True, help="save predicted tracks or not")
parser.add_argument("--save_dir", type=str, default='./saved_tracks_best', help="where enhanced tracks to be saved")
parser.add_argument('--config', default='recipes/SEMamba_advanced/SEMamba_advanced.yaml')

args = parser.parse_args()

with open(args.config) as f:
    data = f.read()

json_config = json.loads(data)
h = AttrDict(json_config)

if __name__ == "__main__":
    noisy_dir = os.path.join(args.test_dir, "noisy")
    clean_dir = os.path.join(args.test_dir, "clean")
    evaluation(args.model_path, noisy_dir, clean_dir, args.save_tracks, args.save_dir, h)
`
![QQ截图20240918202220](https://github.com/user-attachments/assets/7c22e93f-178c-4d37-89a1-0aeceb31d73a)
